src/or/relay.c

```diff
@@ -1783,6 +1782,21 @@ connection_edge_process_relay_cell(cell_t *cell, circuit_t *circ,
                  rh.stream_id);
         return 0;
       }
+      if (conn->package_window + STREAMWINDOW_INCREMENT >
+          STREAMWINDOW_START_MAX) {
+        static struct ratelim_t stream_warn_ratelim = RATELIM_INIT(600);
+        log_fn_ratelim(&stream_warn_ratelim,LOG_PROTOCOL_WARN, LD_PROTOCOL,
+               "Unexpected stream sendme cell. Closing circ (window %d).",
+               conn->package_window);
+        return -END_CIRC_REASON_TORPROTOCOL;
+      }
+
+      /* At this point, the stream sendme is valid */
+      if (CIRCUIT_IS_ORIGIN(circ)) {
+        circuit_read_valid_data(TO_ORIGIN_CIRCUIT(circ),
+                                rh.length);
+      }
```

Comment:
Hmmm I wonder why we need to do this here. Considering that we can make this call a bit earlier within the circuit window check, won't that increment twice the bw counters on the circuit for the same cell?

Reply:
The first call to circuit_read_valid_data() is inside the code block for the !rh.stream_id. That code block handles circuit level SENDMEs (which have stream ID 0), and it returns before the code that handles stream-level SENDMEs executes (which have a stream ID corresponding to the connection they are operating on). This patch only alters the stream-level SENDME branch.